ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsWriter.scala

```diff
@@ -103,23 +134,46 @@ private[dao] object TransactionsWriter extends TransactionsWriter {
       flatTransactionWitnessesBatch.foreach(_.execute())
       complementWitnessesBatch.foreach(_.execute())
 
+      val contractBatches = ContractsTable.prepareBatchInsert(
+        ledgerEffectiveTime = ledgerEffectiveTime,
+        transaction = transaction,
+        divulgedContracts = divulgedContracts,
+      )
+
+      for ((deleted, deleteContractsBatch) <- contractBatches.deletions) {
+        val deleteWitnessesBatch = contractWitnessesTable.prepareBatchDelete(deleted.toSeq)
+        require(deleteWitnessesBatch.nonEmpty, "Illegal: deleting witnesses without a contract")
+        // Delete the witnesses first to respect the foreign key constraint of the underlying storage
+        deleteWitnessesBatch.get.execute()
+        deleteContractsBatch.execute()
+      }
+
+      for ((_, insertContractsBatch) <- contractBatches.insertions) {
+        insertContractsBatch.execute()
+      }
+
+      val notDeleted: ContractId => Boolean =
+        !contractBatches.deletions.fold(Set.empty[ContractId])(_._1).contains(_)
+
+      // Insert the witnesses last to respect the foreign key constraint of the underlying storage.
+      // Compute and insert new witnesses regardless of whether the current transaction adds new
+      // contracts because it may be the case that we are only adding new witnesses to existing
+      // contracts (e.g. via a fetch).
+      val localDivulgence =
+        contractBatches.insertions.fold[WitnessRelation[ContractId]](Map.empty) {
+          case (toBeInserted, _) => divulgence(transaction, blinding.disclosure, toBeInserted)
+        }
+      val fullDivulgence = Relation.union(
+        localDivulgence,
+        blinding.globalDivulgence.filterKeys(notDeleted)
+      )
+      val insertWitnessesBatch = contractWitnessesTable.prepareBatchInsert(fullDivulgence)
+      if (localDivulgence.nonEmpty) {
+        require(insertWitnessesBatch.nonEmpty, "Illegal: inserting a contract without witnesses")
+      }
+      insertWitnessesBatch.foreach(_.execute())
```

Comment:
This is probably a good candidate to be extracted as a method.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/5541/commits/1970436722f3b4966ea9421ec92e24379d737223